lets-revise-hooks-today.md

```diff
@@ -0,0 +1,656 @@
+
+![useEffect.png](https://cdn.hashnode.com/res/hashnode/image/upload/v1609762643281/Yq3P_r__z.png)
+
+In this example we pass data from `App Component` to `ComponentE` use useContext
+
+Let's see how
+
+```
+import React,{useState,createContext} from 'react'
+import ComponenetA from './ComponenetA'
+export const NameContext = createContext()
+
+const AppComponent = () => {
+    const [name,setName]=useState("Minhaj ahmad khan")
+    return (
+        <div>
+            <NameContext.Provider value={'Minhaj ahmad khan'}>
+                <ComponenetA />
+            </NameContext.Provider>
+        </div>
+    )
+}
+
+export default AppComponent
+
+
+```
+
+So here we are export and creating a context `export const NameContext = createContext()` after that we need to provide the context with value.
+The provider must wrap with the children component for the value to be available.
+
+```
+import React,{useContext} from 'react'
+import {NameContext} from './App-Component';
+
+const ComponentE = () => {
+    const userName = useContext(NameContext)
+    return (<div style={{ height: '200px', width: '600px', border: '1px solid red', background: '#e1f5fe' ,margin:'20px'}}>
+        <h1>useContext Example</h1>
+        <h3> hello My Name is  {userName} </h3>
+    </div> )
+}
+
+export default ComponentE
+
+```
+
+Here we are importing the required context after that we are calling `useContext` function and pass in context as its argument which returns the context value
+
+### useReducer
+
+useReducer is a hook that is used for state management. It is an alternative to `useState`.
+
+Similar to useState but gives you more control to manage the state.
+It takes a reducer function and initial state as arguments and returns the state and dispatch method.
+
+Let's see in code which makes more sense
+
+```
+import React, { useReducer } from 'react'
+const initialState = 0;
+const reducer = (state, action) => {
+    switch (action) {
+        case 'increment':
+            return state + 1;
+        case 'decrement':
+            return state - 1;
+        case 'reset':
+            return state = 0;
+        default:
+            return state
+    }
+}
+
+const CounterReducer = () => {
+
+    const [count, dispatch] = useReducer(reducer, initialState)
+    return (
+        <div style={{ height: '200px', width: '600px', border: '1px solid red', background: '#e1f5fe', margin: '20px' }}>
+            <p>Reducer Example</p>
+            <h3>{count}</h3>
+            <button onClick={() => dispatch('increment')}>Increment</button>
+            <button onClick={() => dispatch('decrement')}>Decrement</button>
+            <button onClick={() => dispatch('reset')}>Reset</button>
+
+        </div>
+    )
+}
+
+export default CounterReducer;
+
+```
+
+When you are dealing with multiple state transition it is good to useReducer it makes so easy to maintain state.
+
+It's always a good practice to use `useReducer`when the state depends on the previous one.
```

Comment:
Can you explain this using an example?
When will new state depend on previous state